### PR TITLE
glTFSerializer: Exported textures no longer premultiplied with canvas

### PR DESCRIPTION
**Pull request by kcoley**
- ~~Sets premultipliedAlpha in the playground to false~~
- Exports textures without pre-multiplying the alpha channel with the canvas
- Uses a texture map to avoid duplicating textures with the same uid

### CODE REVIEW

File: serializers/src/glTF/2.0/babylon.glTFMaterial.ts [hunk 1]
```diff
@@ -275,44 +288,69 @@ module BABYLON.GLTF2 {
...
                 if (babylonStandardMaterial.emissiveTexture) {
-                    let promise = _GLTFMaterial._ExportTextureAsync(babylonStandardMaterial.emissiveTexture, mimeType, images, textures, samplers, imageData, useAlpha).then(glTFEmissiveTexture => {
-                        if (glTFEmissiveTexture) {
-                            glTFMaterial.emissiveTexture = glTFEmissiveTexture;
-                        }
-                        glTFMaterial.emissiveFactor = [1.0, 1.0, 1.0];
-                    });
-                    promises.push(promise);
+                    glTFMaterial.emissiveFactor = [1.0, 1.0, 1.0];
+                    if (babylonStandardMaterial.emissiveTexture.uid in this._textureMap) {
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFMaterial.ts]: consider putting `babylonStandardMaterial.emissiveTexture.uid` in a variable

File: serializers/src/glTF/2.0/babylon.glTFMaterial.ts [hunk 2]
```diff
@@ -500,20 +521,64 @@ module BABYLON.GLTF2 {
          * @param mimeType mimetype of the image
          * @returns base64 image string
          */
-        private static _CreateBase64FromCanvas(buffer: Uint8ClampedArray | Float32Array, width: number, height: number, mimeType: ImageMimeType): string {
-            const imageCanvas = document.createElement('canvas');
-            imageCanvas.width = width;
-            imageCanvas.height = height;
-            imageCanvas.id = "WriteCanvas";
+        private static _CreateBase64FromCanvasAsync(buffer: Uint8ClampedArray | Float32Array, width: number, height: number, mimeType: ImageMimeType): Promise<string> {
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFMaterial.ts]: Check if the types for buffer are still correct.

File: serializers/src/glTF/2.0/babylon.glTFMaterial.ts [hunk 3]
```diff
@@ -62,11 +62,25 @@ module BABYLON.GLTF2 {
          */
         private static _maxSpecularPower = 1024;
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFMaterial.ts]: add readonly to constants

File: serializers/src/glTF/2.0/babylon.glTFMaterial.ts [hunk 4]
```diff
@@ -62,11 +62,25 @@ module BABYLON.GLTF2 {
          */
         private static _maxSpecularPower = 1024;
 
+
+        private static _localEngine: Engine;
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFMaterial.ts]: Move this to the exporter as member variables and then either have a parent pointer to the exporter or pass down these values to the material from the exporter.

File: serializers/src/glTF/2.0/babylon.glTFMaterial.ts [hunk 5]
```diff
@@ -500,20 +521,64 @@ module BABYLON.GLTF2 {
...
 
-            imgData.data.set(buffer);
-            ctx.putImageData(imgData, 0, 0);
+                let textureType = Engine.TEXTURETYPE_UNSIGNED_INT;
+                hostingScene = new Scene(this._localEngine);
 
-            return imageCanvas.toDataURL(mimeType);
+                // Create a temporary texture with the texture buffer data
+                let tempTexture = this._localEngine.createRawTexture(buffer, width, height, Engine.TEXTUREFORMAT_RGBA, false, true, Texture.NEAREST_SAMPLINGMODE, null, textureType);
+                let rgbdPostProcess = new PostProcess("pass", "pass", null, null, 1, null, Texture.NEAREST_SAMPLINGMODE, this._localEngine, false, undefined, Engine.TEXTURETYPE_UNSIGNED_INT, undefined, null, false);
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFMaterial.ts]: Dispose this when it's done

File: serializers/src/glTF/2.0/babylon.glTFMaterial.ts [hunk 6]
```diff
@@ -587,10 +652,10 @@ module BABYLON.GLTF2 {
          * @param mimeType the mime type to use for the texture
          * @returns pbr metallic roughness interface or null
          */
-        private static _ConvertSpecularGlossinessTexturesToMetallicRoughness(diffuseTexture: BaseTexture, specularGlossinessTexture: BaseTexture, factors: _IPBRSpecularGlossiness, mimeType: ImageMimeType): Nullable<_IPBRMetallicRoughness> {
+        private static _ConvertSpecularGlossinessTexturesToMetallicRoughnessAsync(diffuseTexture: BaseTexture, specularGlossinessTexture: BaseTexture, factors: _IPBRSpecularGlossiness, mimeType: ImageMimeType): Promise<Nullable<_IPBRMetallicRoughness>> {
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFMaterial.ts]: return type does not need to be Nullable

**Comment by sebavan** [comment]
Is SetAlphaToOne needed ?  it looks like we could do the same by disabling alpha on the colorMask https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/colorMask ?

File: serializers/src/glTF/2.0/babylon.glTFMaterialExporter.ts [hunk 7]
```diff
@@ -500,20 +477,56 @@ module BABYLON.GLTF2 {
...
-            const imageCanvas = document.createElement('canvas');
-            imageCanvas.width = width;
-            imageCanvas.height = height;
-            imageCanvas.id = "WriteCanvas";
-
-            const ctx = imageCanvas.getContext('2d') as CanvasRenderingContext2D;
-
-            const imgData = ctx.createImageData(width, height);
+        private _createBase64FromCanvasAsync(buffer: Uint8Array | Float32Array, width: number, height: number, mimeType: ImageMimeType): Promise<string> {
+            this._createBase64FromCanvasPromiseChain = this._createBase64FromCanvasPromiseChain.then(() => {
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFMaterialExporter.ts]: do you still need to chain the promises?

File: serializers/src/glTF/2.0/babylon.glTFMaterialExporter.ts [hunk 8]
```diff
@@ -1112,34 +1124,46 @@ module BABYLON.GLTF2 {
          * @param imageData map of image file name and data
          * @return glTF texture info, or null if the texture format is not supported
          */
-        private static _ExportTextureAsync(babylonTexture: BaseTexture, mimeType: ImageMimeType, images: IImage[], textures: ITexture[], samplers: ISampler[], imageData: { [fileName: string]: { data: Uint8Array, mimeType: ImageMimeType } }, useAlpha: boolean): Promise<Nullable<ITextureInfo>> {
-            const sampler = _GLTFMaterial._GetGLTFTextureSampler(babylonTexture);
-            let samplerIndex: Nullable<number> = null;
+        private _exportTextureAsync(babylonTexture: BaseTexture, mimeType: ImageMimeType): Promise<Nullable<ITextureInfo>> {
+            return new Promise((resolve, reject) => {
```
Comment (bghgary) [inline serializers/src/glTF/2.0/babylon.glTFMaterialExporter.ts]: this shouldn't be a new promise, use Promise.resolve().then() instead or the errors from _createBase64FromCanvasAsync is not captured.

File: serializers/src/glTF/2.0/babylon.glTFMaterialExporter.ts (hunk 7 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/babylon.glTFMaterialExporter.ts]: Nope, I removed the chained promises.

File: serializers/src/glTF/2.0/babylon.glTFMaterialExporter.ts (hunk 8 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/babylon.glTFMaterialExporter.ts]: Gotcha, changed.

**Pull request merged**